pkg/soc/soc.go

```diff
@@ -22,130 +22,131 @@ const (
-	err := s.AddSigner(signer)
-	if err != nil {
-		return nil, err
+// New creates a new Soc representation from arbitrary soc id and
+// a content-addressed chunk.
+func New(id ID, ch swarm.Chunk) *Soc {
+	return &Soc{
+		id:    id,
+		chunk: ch,
 	}
-	return s.ToChunk()
 }
 
-// NewChunk is a convenience function to create a single-owner chunk ready to be sent
-// on the network.
-func NewSignedChunk(id Id, ch swarm.Chunk, owner, sig []byte) (swarm.Chunk, error) {
+// NewSignedSoc creates a single-owner chunk based on already signed data.
+func NewSigned(id ID, ch swarm.Chunk, owner, sig []byte) (*Soc, error) {
 	s := New(id, ch)
-	s.signature = sig
 	o, err := NewOwner(owner)
 	if err != nil {
 		return nil, err
 	}
 	s.owner = o
+	s.signature = sig
+	return s, nil
+}
 
-	// create chunk
-	socAddress, err := s.Address()
-	if err != nil {
-		return nil, err
+// address returns the soc chunk address.
+func (s *Soc) address() (swarm.Address, error) {
+	if len(s.owner) != crypto.AddressSize {
+		return swarm.ZeroAddress, errNoOwner
 	}
-	return swarm.NewChunk(socAddress, s.toBytes()), nil
+	return CreateAddress(s.id, s.owner)
 }
 
-// New creates a new Soc representation from arbitrary soc id and
-// a content-addressed chunk.
-//
-// By default the span of the soc data is set to the length
-// of the payload.
-func New(id Id, ch swarm.Chunk) *Soc {
-	return &Soc{
-		id:    id,
-		Chunk: ch,
+// WrappedChunk returns the chunk wrapped by the soc.
+func (s *Soc) WrappedChunk() swarm.Chunk {
+	return s.chunk
+}
+
+// Chunk returns the soc chunk.
+func (s *Soc) Chunk() (swarm.Chunk, error) {
+	socAddress, err := s.address()
+	if err != nil {
+		return nil, err
 	}
+	return swarm.NewChunk(socAddress, s.toBytes()), nil
 }
 
-// WithOwnerAddress provides the possibility of setting the ethereum
-// address for the owner of an soc in the absence of a signer.
-func (s *Soc) WithOwnerAddress(ownerAddress *Owner) *Soc {
-	s.owner = ownerAddress
-	return s
+// toBytes is a helper function to convert the RawSoc data to bytes.
+func (s *Soc) toBytes() []byte {
+	buf := bytes.NewBuffer(nil)
+	buf.Write(s.id)
+	buf.Write(s.signature)
+	buf.Write(s.chunk.Data())
+	return buf.Bytes()
 }
 
-// AddSigner currently sets a single signer for the soc.
-//
-// This method will overwrite any value set with WithOwnerAddress with
-// the address derived from the given signer.
-func (s *Soc) AddSigner(signer crypto.Signer) error {
+// Sign signs a soc using the signer.
+// It returns a signed chunk ready for submission to the network.
+func (s *Soc) Sign(signer crypto.Signer) (swarm.Chunk, error) {
+	// create owner
 	publicKey, err := signer.PublicKey()
 	if err != nil {
-		return err
+		return nil, err
 	}
 	ownerAddressBytes, err := crypto.NewEthereumAddress(*publicKey)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	ownerAddress, err := NewOwner(ownerAddressBytes)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	s.signer = signer
 	s.owner = ownerAddress
```

Comment:
I don't think that we need to mutate this field. The same goes for the signature field. If there is no subsequent usage of the `Soc` instance after signing, then this is probably not necessary.

Reply:
The signature is used in the `Chunk` method to create the soc chunk in the return of the `Sign` function, but I can do it in the `Sign` method instead. 

The `owner` is used to generate the soc's address, but it is currently used only in the tests and in the `Valid` method. Removing the owner would require us to extract the signature from the chunk and make the recovery of the owner address in the `Valid` function and use it there to recreate the chunk address for comparison.

Reply:
For example, I could create a signed soc, and then call `Sign` again with a different signer, and if the sign does not mutate the owner, the newly signed soc will have another owner from the signer and result in a different soc address.